File: LibraryCode/CPP/ufds.cpp
/*
Union Find Disjoint sets

- Data structure used to model a collection of disjoint sets which can efficiently determine which set an item belongs to
  as well as to unite two sets

- code taken from cp-book github, here for easy reference

Time: ~O(1) operations
Space: O(N) 
*/
#include <bits/stdc++.h>
using namespace std;

class UnionFind {
private:
    vector<int> p, rank, setSize;
    int numSets;
public:
    UnionFind(int N) {
        p.assign(N, 0); for (int i = 0; i < N; ++i) p[i] = i;
        rank.assign(N, 0);                           // optional speedup
        setSize.assign(N, 1);                        // optional feature
        numSets = N;                                 // optional feature
    }

    int find_set(int i) { return (p[i] == i) ? i : (p[i] = find_set(p[i])); }
    bool is_same_set(int i, int j) { return find_set(i) == find_set(j); }

    int num_disjoint() { return numSets; }      // optional
    int size_of_set(int i) { return setSize[find_set(i)]; } // optional

    void union_set(int i, int j) {
        if (is_same_set(i, j)) return;                 // i and j are in same set
        int x = find_set(i), y = find_set(j);          // find both rep items
        if (rank[x] > rank[y]) swap(x, y);           // keep x 'shorter' than y
        p[x] = y;                                    // set x under y
        if (rank[x] == rank[y]) ++rank[y];           // optional speedup
        setSize[y] += setSize[x];                    // combine set sizes at y
        --numSets;                                   // a union reduces numSets
    }
};

// UnionFind ufds (N);          // Create a UFDS with [0..N-1] items
// ufds.find_set(i);            // returns which set i belongs to
// ufds.is_same_set(i, j);      // returns bool 0/1 if i, j are in same set
// ufds.union_set(i, j);        // combines set containing i and set containing j
// ufds.num_disjoint();         // returns int, number of disjoint sets

File: LibraryCode/JAVA/ufds.java
/**
 * Union Find Disjoint Sets
 * 
 * - Data structure used to model a collection of disjoint sets which can efficiently determine which set an item belongs to
 *  as well as to unite two sets
 * 
 * - code taken from cp-book github, here for easy reference
 * 
 * Time: O(1) operations
 * Space: O(N) 
 */
import java.util.*;
public class Ufds {
    public static class UnionFind {     // OOP style
        private ArrayList<Integer> p, rank, setSize;
        private int numSets;
      
        public UnionFind(int N) {
            p = new ArrayList<>(N);
            rank = new ArrayList<>(N);
            setSize = new ArrayList<>(N);
            numSets = N;
            for (int i = 0; i < N; i++) {
                p.add(i);
                rank.add(0);
                setSize.add(1);
            }
        }
      
        public int findSet(int i) { 
            if (p.get(i) == i) return i;
            else {
                int ret = findSet(p.get(i)); p.set(i, ret);
                return ret; } }
      
        public Boolean isSameSet(int i, int j) { return findSet(i) == findSet(j); }
      
        public void unionSet(int i, int j) { 
            if (!isSameSet(i, j)) { numSets--; 
            int x = findSet(i), y = findSet(j);
            // rank is used to keep the tree short
            if (rank.get(x) > rank.get(y)) { p.set(y, x); setSize.set(x, setSize.get(x) + setSize.get(y)); }
            else                           { p.set(x, y); setSize.set(y, setSize.get(y) + setSize.get(x));
                                            if (rank.get(x) == rank.get(y)) rank.set(y, rank.get(y) + 1); } } }
        public int numDisjointSets() { return numSets; }
        public int sizeOfSet(int i) { return setSize.get(findSet(i)); }
    }
    
    public static void main(String... args) {
        UnionFind uf = new UnionFind(5);
        uf.unionSet(0, 1);
        uf.unionSet(2, 3);
        uf.sizeOfSet(0);              // 2
        uf.isSameSet(1, 3);         // false
    }
}
